lib/Analysis/DebugAnalysis.cpp

```diff
@@ -0,0 +1,129 @@
+//===- DebugAnalysis.cpp --------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Analysis/DebugAnalysis.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/Debug/DebugOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/Support/Debug.h"
+
+using namespace circt;
+using namespace debug;
+using namespace mlir;
+
+namespace {
+struct DebugAnalysisBuilder {
+  DebugAnalysisBuilder(Operation *rootOp) : rootOp(rootOp) {}
+  void run();
+  void addDebugOp(Operation *op);
+  void addDebugValue(Value value);
+  void addDebugOperand(OpOperand *operand);
+  void maybeDebugOp(Operation *op);
+
+  Operation *rootOp;
+  SetVector<Operation *> worklist;
+
+  DenseSet<Operation *> debugOps;
+  DenseSet<Value> debugValues;
+  DenseSet<OpOperand *> debugOperands;
+};
+} // namespace
+
+void DebugAnalysisBuilder::run() {
+  // Find all debug ops nested under the root op and mark them as debug-only
+  // to kickstart the analysis.
+  rootOp->walk([&](Operation *op) {
+    if (isa<debug::DebugDialect>(op->getDialect())) {
+      addDebugOp(op);
+      return;
+    }
+    for (auto &region : op->getRegions())
+      for (auto &block : region)
+        for (auto arg : block.getArguments())
+          if (isa<debug::DebugDialect>(arg.getType().getDialect()))
+            addDebugValue(arg);
+    for (auto result : op->getResults())
+      if (isa<debug::DebugDialect>(result.getType().getDialect()))
+        addDebugValue(result);
+  });
+
+  // Visit operations and check if all their operands or all their uses are
+  // marked as debug-only. If they are, mark the op itself as debug-only.
+  while (!worklist.empty()) {
+    auto *op = worklist.pop_back_val();
+    if (debugOps.contains(op))
+      continue;
+
+    // Do not propagate through stateful elements. This should probably be
+    // configurable, since certain forms of debug info extraction would be able
+    // to pull entire state machines out of the design. For now this just
+    // represents the common denominator across all debug infos.
+    if (!isa<hw::HWDialect, comb::CombDialect>(op->getDialect()))
+      continue;
+    if (op->hasAttr("name"))
```

Comment:
`"name"` attributes, and their CIRCT semantics, should really be defined somewhere - as opposed to having random strings sprinkled through the design.

Reply:
Yeah something like that would be very nice. It would sometimes also be great to just now whether something will have a proper Verilog name in the end, even if you currently don't care about that name.

In this case here the check for "name" is more a stand-in until we flesh out what exactly the debug info assumes to be extractable into some debug file format, and which parts have to stay in Verilog. For example, a debug info file might be able to capture entire registers and state machines, in which case the debug analysis should look through those and eagerly mark more things as debug-only. But other debug files might not be able to even capture simple expressions, so the debug analysis would have to break much earlier. Not entirely sure how we should handle this dependence on the DI file format.